src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -1047,9 +1055,10 @@ void IMergeTreeDataPart::loadColumns(bool require)
     else
     {
         loaded_columns.readText(*volume->getDisk()->readFile(path));
+        loaded_from_disk = true;
     }
 
-    setColumns(loaded_columns);
+    setColumns(loaded_columns, loaded_from_disk);
```

Comment:
I'd prefer to modify columns right here instead of introducing flag in `setColumns`.